Refactor merge constraints on signatures

clementblaudeau opened the pull request:
This PR proposes a refactoring of the `merge_constraint` function, which handles signatures constraints. 

# Context

Constraints are modifiers that act on module types. There are of three forms of constraints :
- **type** constraint `... with type t = ...`
- **module** constraint `... with module X = ...`
- **module type** constraints `... with module type T = ...`

In addition, 

1. Each constraint can be *destructive*, (with the syntax `:=`) meaning that the substituted identifier is removed from the signature. This imposes additional checks to ensure wellformedness in the absence of the destructed item.
2. Each constraint can be *deep*, meaning that the substituted identifier might be inside a submodule (`... with type X.Y.t = ...`, `with module type M.N.T := sig end`, etc). 
3. Each constraint can be *instantiating*, if the substitution replaces an "*abstract*" field (for module constraints, "*abstract*" means "*not already an alias*"). If the constraint is not instantiating, equivalence or subtyping checks with the old definition are performed.
    ```ocaml 
    (sig type t end) with type t = int * int (* instantiating, as [t] was abstract *)
    (sig type t = bool end) with type t = bool (* non-instantiating, the two definitions must be equivalent *) 
    ```
4. Finally, merging is used both in (1) "normal" mode, to build a typed tree, in (2) "approx" mode, during the signature approximation phase of typechecking recursive modules (see #13829), and in (3) "package" mode, for checking signatures of first-class modules.

# Refactoring

The refactoring is limited to `typemod.ml`. It consists of : 
1. Creating a `Merge` module to gather the merge related functions
2. Extract the mutually recursive functions `patch_deep_item` and `merge_signature` to handle deep constraints
3. Create separated (non-recursive) functions `merge_type`, `merge_module` and `merge_modtype`.
4. Add documentation comments

Now, the structure is similar for each form of merging:
1. A `patch` function is defined to identify the item to substitute, do the equivalence checks if needed, and optionally build the new replacement item (if the substitution is non-destructive)
2. The patch is applied to the module type by using the general `merge` function. It returns the path of the affected item (along with the list of all suffixes, if the substitution was deep).
3. Some post processing is applied (actual replacement for destructive substitutions, wellformedness checks) 

# ~~Semantic changes~~

[Removed, moved to #13929]

The only visible change is the introduction of a new error when doing deep destructive substitutions inside aliases signatures, as in : 
```ocaml
module X = sig type t = int val x : t end 
module type T = sig module Y = X end with type Y.t := int
```
Throws a new error `With_lost_alias` : 
 ```
 Error: This deep destructive "with" substitution inside of "Y" would
       loose the aliasing to "X" .
```

samsa1 commented:
I now this might be a bit complicated but could it be possible to split this PR into multiple smaller commits ? As is in the diff it is quite complicated to identify which parts of the code have been rewritten and which have just been moved around.

This could help reviewing by clearly identifying when code is just moved around.
For example lifting `unsafe_signature_subst` out of the function it was defined in previously could be as single commit.

clementblaudeau committed:
[refactor merge] Create a Merge module + white space changes

clementblaudeau committed:
[refactor merge] Extracted helper functions

- [split_row_ids] and [unsafe_signature_subst]

clementblaudeau committed:
[refactor merge] Changed the return logic

- Typedtrees are no longer built inside [merge_constraint] but inside
[transl_with], which removes the need for a special approximation case: merging
always returns a module type, not a Typedtree. Changed [transl_with] to build
the Typedtree there.

- Removed the [real_ids] mechanism that was used to store (imperatively) the
list of affected paths. Now the patches (defined by [return], [return_payload]
and [return_paths]) store both the resulting path and the list of affected
paths.

- Added a [payload] mechanism used only for the type constraint case, where the
replacement declaration is returned as an additional payload. Other cases return
[None].

- Create separate functions ([merge_type], [merge_module], [merge_modtype])

- Extracted the post processing (wellformedness checks and substitutions) into
an helper function [post_process]

clementblaudeau committed:
[refactor merge] Extracted the recursive functions for deep constr

- Broke down the main merging function into three parts: [merge_signature],
[patch_deep_item] and [patch_all]. The first two are mutually recursive and use
an extra argument [~patch]. For now, the only patch provided (in [merge_type],
[merge_module], etc) is [patch_all].

clementblaudeau committed:
[refactor merge] Moved code

clementblaudeau committed:
[refactor merge] Extracted the type-package case

- Move the patching logic from [patch_all] to a specialized patch function in
[merge_package_constraint]

- Removed the (now) useless lid argument for [patch_all]

clementblaudeau committed:
[refactor merge] Extracted the type case

- Moved the patching logic from [patch_all] to specialized patch function in
[merge_type]

- Removed the (now) useless [initial_env] argument for [patch_all]

clementblaudeau committed:
[refactor merge] Extracted the module case

- Moved the patching logic from [patch_all] to a specialized patch function in
[merge_module]. Merged the common parts of the destructive and non-destructive
cases

clementblaudeau committed:
[refactor merge] Extracted the modtype case

- Moved the patching logic from [patch_all] to a specialized patch function in
[merge_modtype]

- Removed the (now) useless [patch_all]

clementblaudeau committed:
[refactor merge] Added comments, minor stylistic changes

- Added general comments about merging

- Renamed the `merge_package_constraint` into `merge_package`

- Moved the [try ... with] of [merge_signature] to the outer function [merge].

clementblaudeau commented:
I've split the PR into smaller commits to simplify the review. The first commit should be reviewed without the whitespace changes, as it indents all the affected code. In my opinion, the only commit that remains tricky is [5c345a2](https://github.com/ocaml/ocaml/pull/13911/commits/5c345a20690c98a9cd7dff23f2d8e5c5c33e7596), where the logic for what is returned by `merge_constraint` is changed. I've documented in each commit what is going on, but I can provide more info if needed for reviewing.

samsa1 commented on typing/typemod.ml:
```diff
@@ -503,305 +510,292 @@ let merge_constraint_aux initial_env loc sg lid constr : merge_result =
-    (* When the constraint affects a component of a submodule *)
-    | Sig_module(id, _, md, rs, priv) as current_item, s :: namelist, _
+    with Includemod.Error explanation ->
+      raise(Error(loc, initial_env, With_mismatch(lid.txt, explanation)))
```
This `try ... with ...` is used at every recursive call instead of only once in the previous version.
As `merge_signature` is called only by `merge` outside of this definition it might be cleaner to move it there.

samsa1 commented on Changes:
```diff
@@ -371,6 +371,10 @@ Working version
 - #13818: better delimited hints in error message
   (Florian Angeletti, review by Gabriel Scherer)
 
+- #13911: Added an `Lost alias` error message for deep destructive substitutions
```
This change might break a few codes due to the new error message. Thus you should use a `*`

samsa1 reviewed (approved):
As a whole this PR significantly increases the readability of the code.
We previously had only one big recursive function that handled lots of cases. Now we only have one small recursive function that is parametrized by a patching function that does only what we want it to do.

Thus future maintenance of this code seems much simpler as different cases seems less likely to interact together.

Octachron commented on testsuite/tests/typing-modules/merge_constraint.ml:
```diff
@@ -422,3 +422,44 @@ Error: In this "with" constraint, the new definition of "M.N"
+                      ^^^^^^^^^^^^^^^^^^^^^^^
+Error: This deep destructive "with" substitution inside of "Y" would
+       loose the aliasing to "X" .
+|}]
```
The testsuite should pass, with the current result. A failing testsuite can make harder to bisect issues.

Octachron commented on typing/typemod.ml:
```diff
@@ -3414,6 +3415,14 @@ let report_error ~loc _env = function
+      Location.errorf ~loc
+        "@[<v>\
+           @[This deep destructive %a substitution inside of %a would @ \
+             loose the aliasing to %a @].@]"
```
loose ⇒ lose. Also it is better to avoid opening boxes when possible, and add break hints  before and after large inline block of code:
```ocaml
This deep destructive %a substitution inside@ of@ %a@ would @ \
lose the aliasing to %a.
```

clementblaudeau commented:
I've removed the semantic change of the introduction of an error to #13299. I've force-pushed to a new history where the introduction of the error case and tests is removed, the rest should be unchanged. 
The failing CI comes from the untouched Changelog, but I don't think purely refactoring prs (which this one now is) need an entry in the changelog right ?

Octachron commented on typing/typemod.ml:
```diff
@@ -470,6 +470,27 @@ and merge_info =
+    in
+    split [] ghosts
+
+  let unsafe_signature_subst sub sg loc initial_env =
```
It generally helps to keep the same order of argument across related functions: 
 ```ocaml
 let unsafe_signature_subst initial_env loc sg sub
 ```

clementblaudeau replied:
Fixed in the latest commits

Octachron commented on typing/typemod.ml:
```diff
@@ -447,29 +447,28 @@ type merge_constraint =
+    Some ((path, [path], None), {Signature_group.ghosts; replace_by})
+
+  let return_payload  ~payload ~ghosts ~replace_by path =
+    Some ((path, [path], payload), {Signature_group.ghosts; replace_by})
```
This could be an optional argument, and probably named `late_typedtree` to be more explicit on the use case of this payload.

Octachron commented on typing/typemod.ml:
```diff
@@ -447,29 +447,28 @@ type merge_constraint =
-     (without typedtree) *)
-  | No_TypedTree
+type merge_result = Path.t * (Path.t list)
+                    * (Typedtree.type_declaration option) * Types.signature
```
This could be a record.

Octachron commented on typing/typemod.ml:
```diff
@@ -491,37 +490,45 @@ module Merge = struct
+      else sg
+    in
+    (* check that the resulting signature is still wellformed *)
+    let _ = check_well_formed_module env loc "this instantiated signature"
```
```
let () = ...
```
to avoid suggesting that `check_well_formed_module` returns some ignored contents.

Octachron commented on typing/typemod.ml:
```diff
@@ -660,16 +651,41 @@ module Merge = struct
+        | Some ((p, paths, payload), sg) -> p, paths, payload, sg
+        | None -> raise(Error(loc, initial_env, With_no_component lid.txt))
+      end
+    with Includemod.Error explanation ->
```
I would propose to merge the `try ... with` and the `match .... with ` into one `match ... with `.

clementblaudeau replied:
The catching of the error has latter been moved to the `merge` function

Octachron commented on typing/typemod.ml:
```diff
@@ -508,15 +505,9 @@ module Merge = struct
-    let rec patch_item constr namelist outer_sig_env sg_for_env ~ghosts item =
+    (* broken indentation to limit diff *)
+  let patch_all ~destructive ?(approx=false) initial_env loc lid constr =
+    fun item s outer_sig_env sg_for_env ~ghosts ->
```
This is an uncommon styles. You can simply split the list of parameters.

clementblaudeau committed:
[refactor merge] Corrections from Octachron review

clementblaudeau committed:
[refactor merge] Moved the catching of exception

clementblaudeau committed:
[refactor merge] Updated comments

Octachron reviewed (approved):
I agree that sharing only the recursive submodule traversal between the various `with` cases makes for simpler code. The code is ready to merge once the conflict is resolved.

clementblaudeau commented:
I've rebased the branch on trunk.

Octachron commented:
You are missing a Changes entry (inside the "Internal/compiler-libs changes" section).

clementblaudeau committed:
[refactor merge] Added Changelog entry
